### PR TITLE
Add new package: ffte

### PR DESCRIPTION
**Pull request by luszczek**

### CODE REVIEW

**Comment by bernhardkaindl** [comment]
@luszczek The style chec failed. See the log of the check, but even better do this:

Before git push or creation of the merge request, please generally always run:
```bash
p=ffte  ; spack style var/spack/repos/builtin/packages/$p/package.py
```
If `spack style` has an error bootstrapping `isort`, `flake8` or `mypy`, install these using `pip3`:
```yaml
pip3 install isort
pip3 install flake8
pip3 install mypy
```

**Review by bernhardkaindl — CHANGES_REQUESTED** [review]
@luszczek The style check failed. See the log of the check:

```patch
   @@ -3,8 +3,9 @@
   #
   # SPDX-License-Identifier: (Apache-2.0 OR MIT)
  
  +import glob
  +
   from spack import *
  -import glob
  ```
Do this:

After applying the fix, please check that the style check succeeds then on your machine befor pushing the update to the github branch of this PR:

Before git push or creation of the merge request, please generally always run:

p=ffte  ; spack style var/spack/repos/builtin/packages/$p/package.py

If spack style has an error bootstrapping isort, flake8 or mypy, install these using pip3:

pip3 install isort
pip3 install flake8
pip3 install mypy

**Comment by luszczek** [comment]
Thank you for the suggestion.I ran `spack style` but it failed for every package it tried to install. And I don't control the system's Python or PIP packages. Spack insists to install it's own copies of everything instead of just accept externally installed packages. I'm happy to run the checks manually but it's hard to find the exact invocation in `lib/spack/spack/cmd/style.py`.

I fixed what I could find with my setup.

**Comment by G-Ragghianti** [comment]
> Thank you for the suggestion.I ran `spack style` but it failed for every package it tried to install. And I don't control the system's Python or PIP packages. Spack insists to install it's own copies of everything instead of just accept externally installed packages. I'm happy to run the checks manually but it's hard to find the exact invocation in `lib/spack/spack/cmd/style.py`.
> 
> I fixed what I could find with my setup.

I've found that I need to do "spack install py-isort; spack load py-isort" before I can use the style checks.  I'm doing this now after I test a basic install of the package.

**Comment by luszczek** [comment]
Just for completeness, this is the output from the `style` command:
```
==> Running style checks on spack
  selected: isort, mypy, flake8
==> Modified files
  var/spack/repos/builtin/packages/ffte/package.py
==> Running isort checks
  Traceback (most recent call last):
    File "/home/luszczek/.spack/bootstrap/store/linux-centos7-zen/gcc-7.3.0/py-isort-5.9.1-67frdjudcqvjcpp442imn7kxeslicpqf/bin/isort", line 6, in <module>
      from pkg_resources import load_entry_point
    File "/usr/lib/python3.6/site-packages/pkg_resources/__init__.py", line 3095, in <module>
      @_call_aside
    File "/usr/lib/python3.6/site-packages/pkg_resources/__init__.py", line 3079, in _call_aside
      f(*args, **kwargs)
    File "/usr/lib/python3.6/site-packages/pkg_resources/__init__.py", line 3108, in _initialize_master_working_set
      working_set = WorkingSet._build_master()
    File "/usr/lib/python3.6/site-packages/pkg_resources/__init__.py", line 570, in _build_master
      ws.require(__requires__)
    File "/usr/lib/python3.6/site-packages/pkg_resources/__init__.py", line 888, in require
      needed = self.resolve(parse_requirements(requirements))
    File "/usr/lib/python3.6/site-packages/pkg_resources/__init__.py", line 774, in resolve
      raise DistributionNotFound(req, requirers)
  pkg_resources.DistributionNotFound: The 'isort==5.9.1' distribution was not found and is required by the application
  isort found errors
==> Running mypy checks
==> Bootstrapping clingo from pre-built binaries
[+] /home/luszczek/workspace/pyenv/withscipy (external python-3.8-cyouubtf7nvntuxiltnfvmgw5bm6rwum)
==> Installing py-setuptools-57.4.0-mphk4k7rzacqx5gqewwwovqsjkeciesh
==> No binary for py-setuptools-57.4.0-mphk4k7rzacqx5gqewwwovqsjkeciesh found: installing from source
==> Fetching https://mirror.spack.io/_source-cache/archive/6b/6bac238ffdf24e8806c61440e755192470352850f3419a52f26ffe0a1a64f465.tar.gz
==> No patches needed for py-setuptools
==> py-setuptools: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 1:
    '/home/luszczek/workspace/pyenv/withscipy/bin/python3' '-s' 'setup.py' '--no-user-cfg' 'build'

1 error found in build log:
     17      sys.path = [
     18        '/home/luszczek/.spack/bootstrap/store/linux-centos7-zen/gcc-7.3.0/py-setuptools-57.4.0-mphk4k7rzacqx5gqewwwovqsjkeciesh/lib/python3.8/site-packages',
     19        '/home/luszczek/workspace/pyenv/withscipy/lib/python38.zip',
     20        '/home/luszczek/workspace/pyenv/withscipy/lib/python3.8',
     21        '/home/luszczek/workspace/pyenv/withscipy/lib/python3.8/lib-dynload',
     22      ]
  >> 23    Fatal Python error: init_fs_encoding: failed to get the Python codec of the filesystem encoding
     24    Python runtime state: core initialized
     25    ModuleNotFoundError: No module named 'encodings'
     26
     27    Current thread 0x00007fc8960a5740 (most recent call first):
     28    <no Python frame>

See build log for details:
  /tmp/luszczek/spack-stage/spack-stage-py-setuptools-57.4.0-mphk4k7rzacqx5gqewwwovqsjkeciesh/spack-build-out.txt

==> Warning: Skipping build of py-toml-0.10.2-o27kfmebyunobctlqtjlbp5tuwly6lr7 since py-setuptools-57.4.0-mphk4k7rzacqx5gqewwwovqsjkeciesh failed
==> Warning: Skipping build of py-mypy-0.910-wcxolsmfe67oefiv6uyt6ksyo6ov3clk since py-toml-0.10.2-o27kfmebyunobctlqtjlbp5tuwly6lr7 failed
==> Warning: Skipping build of py-typed-ast-1.4.3-fnekbn3ddvxu5mydchr6joldtv3xgizc since py-setuptools-57.4.0-mphk4k7rzacqx5gqewwwovqsjkeciesh failed
==> Warning: Skipping build of py-typing-extensions-3.10.0.0-esm6uo5hlfjnauwhpk2p5gjx4gglfn4b since py-setuptools-57.4.0-mphk4k7rzacqx5gqewwwovqsjkeciesh failed
==> Warning: Skipping build of py-mypy-extensions-0.4.3-jl6qlhmlgb2bypatl23ciz6q7r7y2frk since py-setuptools-57.4.0-mphk4k7rzacqx5gqewwwovqsjkeciesh failed
==> Error: py-mypy-0.910-wcxolsmfe67oefiv6uyt6ksyo6ov3clk: Package was not installed
==> Error: Couldn't bootstrap mypy:
Installation request failed.  Refer to reported errors for failing package(s).
```

**Comment by bernhardkaindl** [comment]
@luszczek That looks familiar to me, I think I had the same error initially on the server that I have no control on to install packages.
There pip also isn't installed and I think saw the same error of `ModuleNotFoundError: No module named 'encodings'`. I think a python installation without the python `'encodings'` module might what I causing this failure.

I was able to work around this by instally `pip` into my local home directory using:
$ wget https://bootstrap.pypa.io/get-pip.py *or*
$ curl -sSL https://bootstrap.pypa.io/get-pip.py -o get-pip.py
$ python get-pip.py
Then you ensure that `$HOME/.local/bin` is in your `PATH` and you can use your local `pip` tool without touching the system itself.
Does this help?

PS: You also need to remove the boilerplate at the to of the file, see this error of the failed check:
```py
E Right contains more items, first extra item:
'/home/runner/work/spack/spack/var/spack/repos/builtin/packages/ffte/package.py:18: boilerplate needs to be removed:
# FIXME: Add proper versions and checksums here.'
```
Ensure that the first lines of your `package.py` file has exactly the same header as the other `package.py` files in the repo (you can compare the files using `diff a b | head`)

File: var/spack/repos/builtin/packages/ffte/package.py
Comment (G-Ragghianti) [inline var/spack/repos/builtin/packages/ffte/package.py]: Typo here "Trnasform"

File: var/spack/repos/builtin/packages/ffte/package.py
Comment (G-Ragghianti) [inline var/spack/repos/builtin/packages/ffte/package.py]: Can remove this

**Review by G-Ragghianti — CHANGES_REQUESTED** [review]

**Comment by G-Ragghianti** [comment]
I've tested most of the functionality.  Two thoughts:

1. cuda variant did not work for me.  What do I need to do to get that working?  If it doesn't currently work without some other modification, then I recommend that we leave off the cuda variant.
2. I found that it does not compile with gcc version 10 and greater.  I recommend putting a `conflicts('%gcc@10:')` for that.

**Comment by luszczek** [comment]
> I've tested most of the functionality. Two thoughts:
> 
>     1. cuda variant did not work for me.  What do I need to do to get that working?  If it doesn't currently work without some other modification, then I recommend that we leave off the cuda variant.
> 
>     2. I found that it does not compile with gcc version 10 and greater.  I recommend putting a `conflicts('%gcc@10:')` for that.

Both issues are now fixed.

File: var/spack/repos/builtin/packages/ffte/package.py
Comment (bernhardkaindl) [reply var/spack/repos/builtin/packages/ffte/package.py]: Fixed.

**Review by bernhardkaindl — DISMISSED** [review]
Looks good, all my tests passed, also with +cuda with the Nvidia compiler installed `spack install nvhpc ; spack load nvhpc` loaded, and found by spack using `spack compiler find`

Does it make sense to really provide all the old versions as well (just a question, I have really no idea!)

Secondly, does it make sense to add the version where 
FFT kernels are generated by SPIRAL ?
```
version('7.0-spiral', sha256='e13215b3bedcb910f7efca2d576fb3d3e0fb51675603f6ee5409d6fac44699b9')
```
I guess the variant `spiral` does the same as that version.

This is an approval, a ready to merge from me.

I just wanted to raise two items for discussion. For me it would be enough just to have the latest version initially as there are no existing users. Later, the general plan is to not remove existing checksums, but in spack, if the package and source supports it, all versions can still be built even if no checksum is present for the version. The decision is on you, if you want to start off the new package already with 7 versions. ffte is tiny and quick to compile. But to verify all the verrsions with all compilers would require quite some scripting and takes a bit of time.

@luszczek  Usually I'd try to not start off with a large number of combinations which would be "supported" in all future (including new compilers etc). I'd normally like to have the least number of uncertainities as possible.

@G-Ragghianti - and your review ?

**Comment by luszczek** [comment]
I addressed the issues and pushed the changes.

Here is the script I used to test the versions and variants:
```
for v in `seq 1 7`; do
    spack install ffte@${v}.0
    for t in %gcc@10.2.0 +mpi +vector +mpi+vector; do
        spack install ffte@${v}.0$t
    done
done
```

I don't see any other issues that need addressing.

**Review by bernhardkaindl — APPROVED** [review]

File: var/spack/repos/builtin/packages/ffte/package.py
Comment (bernhardkaindl) [inline var/spack/repos/builtin/packages/ffte/package.py]: @luszczek  A check failed on that, we can't merge without this fixed:
SyntaxError: invalid syntax in /home/spack-test/spack/var/spack/repos/builtin/packages/ffte/package.py, line 13

Use `"""` instead of `'` for the class and method documentation strings (like before).
Running flake8 as style check on your change should be enough the syntax after changing to `"""`.

**Review by bernhardkaindl — CHANGES_REQUESTED** [review]

**Review by bernhardkaindl — APPROVED** [review]

**Review by G-Ragghianti — APPROVED** [review]

**Pull request merged**
